src/qml/qmlautoreload.cpp

```diff
@@ -18,25 +18,40 @@ QmlAutoReload::QmlAutoReload() {
 }
 
 QUrl QmlAutoReload::intercept(const QUrl& url, QQmlAbstractUrlInterceptor::DataType type) {
-    if (!url.isLocalFile() || !QFileInfo(url.toLocalFile()).isFile()) {
+    const auto path = url.toLocalFile();
+
+    if (!url.isLocalFile() || !QFileInfo(path).isFile()) {
+        return url;
+    }
+
+    // Do not add path if it's already watched. See note in the
+    // `QFileSystemWatcher::fileChanged()` documentation.
+    if (m_fileWatcher.files().contains(path)) {
         return url;
     }
-    m_fileWatcher.addPath(url.toLocalFile());
+
+    m_fileWatcher.addPath(path);
     return url;
 }
 
 void QmlAutoReload::slotFileChanged(const QString& changedFile) {
     qDebug() << "File" << changedFile << "used in QML interface has been changed.";
+
     // This is to prevent double-reload when a file is updated twice
-    // in a row as part of the normal saving process. See note in
-    // QFileSystemWatcher::fileChanged documentation.
-    if (m_fileWatcher.removePath(changedFile)) {
-        emit triggered();
-    }
+    // in a row as part of the normal saving process. See note in the
+    // `QFileSystemWatcher::fileChanged()` documentation.
+    //
+    // Unfortunately we cannot be sure that `QFileSystemWatcher::removePath()`
+    // actually returns true, so we need to emit the signal anyway.
+    m_fileWatcher.removePath(changedFile);
+    emit triggered();
```

Comment:
doesn't this unconditional trigger now result in the double-reload?

Reply:
Yes, if the editor writes the file twice.
The idea was to remove the path from the watcher, and once the path is added back (when QML was reloaded) the second write already happened. 

However, `removePath` may return false for many reasons and we cannot rely on it. On my system, if I edit a watched QML file in vim and then save it with `:w` then for some reason the method returns false and no reload is triggered. I don't know why, but this makes autoreload useless for me.

Is double reload really an issue? At worst it's a small optical disturbance and can be expected. On the other hand, the feature is currently broken for me. Does it work for you?

Reply:
No not a problem. I was just questioning my understanding of the code. It no longer matches the comment where it says that it is supposed to prevent the double-reload, so can you remove it? If the double-reload is annoying, we could try to workaround it by implementing some debouncing logic I guess, but we can do that once we deem it to be an actual problem.

Reply:
`m_fileWatcher.removePath` should return `true` for the signal to be emitted. If it return false, it means the file isn't watched anymore, which happens when a signal has already `triggered` a reload. This is the suggested implementation as per the [Qt doc](https://doc.qt.io/qt-6/qfilesystemwatcher.html#fileChanged)

> Note: As a safety measure, many applications save an open file by writing a new file and then deleting the old one. In your slot function, you can check watcher.files().contains(path). If it returns false, check whether the file still exists and then call addPath() to continue watching it.

Reply:
> `m_fileWatcher.removePath` should return `true` for the signal to be emitted. If it return false, it means the file isn't watched anymore, which happens when a signal has already `triggered` a reload.

As I explained above, this is not correct. If you look at the documentation of [`removePath`](https://doc.qt.io/qt-6/qfilesystemwatcher.html#removePath):

> Reasons for watch removal failing are generally system-dependent, but may be due to the path having already been deleted, for example.

If I add logging immediately before the `removePath` call, I can see that the slot is called *exactly once* and that it fails:

    warning [Main] m_fileWatcher.removePath "/home/jan/Projects/mixxx/res/qml/Knob.qml" - result false

So if my editor deleted the file when saving and wrote a new one, it's entirely correct that `removePath` returns false because the old file was deleted, even though triggered was never called before. And this is just an example, I have no idea how my editor is implemented and why `removePath` returns false.

Reply:
> If I add logging immediately before the `removePath` call, I can see that the slot is called _exactly once_ and that `removePath` returns false

This behavior is 100% reproducible with neovim, gedit and `sed -i` which are all text editors that I have installed. Only `touch Knob.qml` will result in a `true` return value. However, with the fixes from this PR it makes no difference. In all cases the QML GUI is reloaded.

Can you give an example of an editor that benefits from checking the return value? What happens if we don't? Won't it still work, just auto-reload twice?

Reply:
> Is double reload an actual problem? Which editors update the file twice in row? If there are any, then it's a small optical disturbance at worst. On the other hand, the feature is currently broken for me. Does it work for you?

I know VS Code relies on this append-delete file edit policy for resilience (whereas vim relies on a swap file along side the edited file if I recall properly?). The disturbance isn't neglectable if you are using some i3-style windows manager. 

I have managed to reproduce it locally to using vim, so I'll see if I can find a way to get both saving policies to work nicely.

Reply:
I have worked on a fix that ensure single reload for both policies [here](https://github.com/mixxxdj/mixxx/compare/main...acolombier:mixxx:fix/qml-auto-reload), could you confirm it works for you?

If yes, perhaps, you can get it in your PR, since you've also added some code cleanup that makes sense to get in (`QmlAutoReload::clear`, var for `url.toLocalFile()`)

Reply:
Have you had a chance to look at the patch above @Holzhaus ?

Reply:
yes, works for me.